Add support for cake qdisc

hnyman opened the pull request:
There was some discussion two weeks ago that cake qdisc might be added to LEDE sources. 
(see https://patchwork.ozlabs.org/patch/621917/ )

Below is my proposal for an easy addition of cake:

* Add 'cake' qdisc itself
  - Use version from 2016-05-19 with the compilation fix for kernel 4.4+ by @kdarbyshirebryant 
    (The later modification for detection of early kernel 4.5.x versions is not needed as they do not exist in LEDE and will not be introduced. Current versions of 4.4, 4.5 and 4.6 are identical for necessary part)
  - Leave the newer cake modifications (heap changes, cobalt etc.) out for now as they seem to be under active development. Wait until cake devs are happy with them.
* Add cake support to 'tc' in iproute2
  - Use a patch to modify tc instead of adding a new tc-adv package from @dtaht 's repo.
    Patch creates q_cake.c that matches commit https://github.com/dtaht/tc-adv/commit/75d8fc53f589f06afe2fc563b7036d9412392a3a with the exception a few column header spacings that have been beautified from the originals in https://github.com/dtaht/tc-adv/commit/75d8fc53f589f06afe2fc563b7036d9412392a3a .
  - Do not include the other things from tc-adv (cake0, cake2, pie etc.).

I made the PR to add the cake package to LEDE itself instead of the packages feed, as there are changes to iproute2 and to ensure that the kernel package gets into the correct download repo.

If the package needs a formal maintainer, I suggest that it would be one of cake devs, e.g. @kdarbyshirebryant 

After cake has been compiled for LEDE trunk, I think that SQM-scripts may need modifications / new cake scripts in order to fully utilise the newest cake modes. I am not sure if layer_cake or piece_of_cake use all the new functionalities. But in any case, testing and development will be easier after cake is compiled by buildbot for various platforms. (cc also to @moeller0 and @tohojo)

Cake compiles for ar71xx/WNDR3700 with this. I have been testing it in my WNDR370/3800 builds for some months.
```
Linux lede 4.4.11 #2 Tue May 31 08:43:20 UTC 2016 mips GNU/Linux
LEDE Reboot r462 / 2016-05-31 11:41 / 1205e9a mediatek: more nand fixes


root@lede:~# tc -s -d qdisc show | head -n 31
qdisc noqueue 0: dev lo root refcnt 2
 Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0)
 backlog 0b 0p requeues 0
qdisc fq_codel 0: dev eth0 root refcnt 2 limit 10240p flows 1024 quantum 1514 target 5.0ms interval 100.0ms ecn
 Sent 2736659712 bytes 4003690 pkt (dropped 0, overlimits 0 requeues 3)
 backlog 0b 0p requeues 3
  maxpacket 1514 drop_overlimit 0 new_flow_count 4 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc cake 800d: dev eth1 root refcnt 2 bandwidth 10Mbit diffserv4 flows rtt 100.0ms raw
 Sent 4342061029 bytes 4344138 pkt (dropped 147202, overlimits 9825484 requeues 0)
 backlog 0b 0p requeues 0
 memory used: 501952b of 500000b
 capacity estimate: 10Mbit
                 Bulk  Best Effort       Video       Voice
  thresh        10Mbit    9375Kbit    7500Kbit    2500Kbit
  target         5.0ms       5.0ms       5.0ms       7.3ms
  interval     100.0ms     100.0ms     100.0ms     102.3ms
  pk_delay       6.5ms       5.0ms       1.1ms       1.2ms
  av_delay       1.9ms       1.6ms       260us       312us
  sp_delay        15us        31us        18us        10us
  pkts         1419271     3069332         166        2571
  bytes      970162736  3513032347       14940      335534
  way_inds      204586       34940           0           0
  way_miss       42881       13278         166         160
  way_cols           0           0           0           0
  drops         107108       40094           0           0
  marks              0           0           0           0
  sp_flows           0           0           0           0
  bk_flows           0           1           0           0
  last_len         115        1494          90         122
  max_len         1514        1514          90         590
```

tohojo commented:
I agree that patching tc is better than messing around with the tc-adv package. However, I would check with Jonathan (@chromi) whether we're ready to freeze the userspace API first.

No opinion on whether the kmod- package should be here or in packages. Theoretically the iproute2 patches can live independently of the package (given a userspace API freeze); I'm sure the people eventually approving this PR will have an opinion on that :)

chromi commented:
> On 31 May, 2016, at 14:34, Toke Høiland-Jørgensen <notifications@github.com> wrote:
> 
> I agree that patching tc is better than messing around with the tc-adv package. However, I would check with Jonathan (@chromi) whether we're ready to freeze the userspace API first.

The userspace API should by now be forwards compatible (not the same as “frozen”), with the possible exception of a particular misfeature that should be deleted.

There is also the question of user documentation of the shortcut keywords for specifying packet overhead compensation.  AFAIK that is still absent from the tc-cake manpage.

I could work on those two points tomorrow.

The sch_cake kmod itself is currently at a stable point immediately before a period of planed experimentation.  I recommend importing specifically commit adddf1056896f970166933d390d2300c5a00cd08 (which is HEAD right now) to preserve that.

 - Jonathan Morton

moeller0 commented:
Hi Jonathan,

> On May 31, 2016, at 14:50 , chromi <notifications@github.com> wrote:
> 
> 
> > On 31 May, 2016, at 14:34, Toke Høiland-Jørgensen <notifications@github.com> wrote:
> > 
> > I agree that patching tc is better than messing around with the tc-adv package. However, I would check with Jonathan (@chromi) whether we're ready to freeze the userspace API first.
> 
> The userspace API should by now be forwards compatible (not the same as “frozen”), with the possible exception of a particular misfeature that should be deleted.

	You are talking about wash here? In the spirit of getting cake upstream, lrt’s just remove that, we could always add it later if someone besides me actualy can present a reasonable use case ;)

> 
> There is also the question of user documentation of the shortcut keywords for specifying packet overhead compensation. AFAIK that is still absent from the tc-cake manpage.

	I would prefer very much if we could just rip the keywords out and just leave “overhead N” in there. As is the keywords are inconsistent in that some act as modifiers that “add” to already specified values while others replace older values whole-sale. Also these keywords do not really simplify the challenges in choosing the correct overhead at all, so my vote would be to go from today’s:

      [ atm | noatm* ] [ overhead N | conservative | raw* ]

to:
	[ atm | noatm* ] [ overhead N | raw* ]

	So, instead of documenting the keywords, remove even more (and add an info/error message to tc to warn users still using the old ones). If that is not acceptable then I would propose to simply create named keywords for the individual components that make up the overhead and apply these in an additive fashion, like:

	ppp pppoe ethernet llc snap atmpad aal5sar

which would translate into:

	2 + 6 + 14 + 3 + 5 + 2 + 8 = 40

	So, I propose to just take the components and their names from https://web.archive.org/web/20150606220856/http://ace-host.stuart.id.au/russell/files/tc/tc-atm/ and ad vlan, mac, and fcs to the mix (and maybe the effective additional overhead required to shape ethernet, IFG and friends). 

	But honestly I believe we would be better off by directing users to decent information how to deduce the applicable overhead in each individual case. Say the link above and maybe a link to https://github.com/moeller0/ATM_overhead_detector might be of more help than trying to simply a complex situation?
	While at it, I also would remove the “conservative” keyword since with that we make a promise to our users we might not be able to keep as we have no guarantee of a maximum overhead possible.

Best Regards
	Sebastian

> 
> I could work on those two points tomorrow.
> 
> The sch_cake kmod itself is currently at a stable point immediately before a period of planed experimentation. I recommend importing specifically commit adddf1056896f970166933d390d2300c5a00cd08 (which is HEAD right now) to preserve that.
> 
> - Jonathan Morton
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

moeller0 commented:
Dear All,

“small correction”


> On May 31, 2016, at 15:23 , moeller0 <moeller0@gmx.de> wrote:
> 
> Hi Jonathan,
> 
>> On May 31, 2016, at 14:50 , chromi <notifications@github.com> wrote:
>> 
>> 
>>> On 31 May, 2016, at 14:34, Toke Høiland-Jørgensen <notifications@github.com> wrote:
>>> 
>>> I agree that patching tc is better than messing around with the tc-adv package. However, I would check with Jonathan (@chromi) whether we're ready to freeze the userspace API first.
>> 
>> The userspace API should by now be forwards compatible (not the same as “frozen”), with the possible exception of a particular misfeature that should be deleted.
> 
> 	You are talking about wash here? In the spirit of getting cake upstream, lrt’s just remove that, we could always add it later if someone besides me actualy can present a reasonable use case ;)
> 
>> 
>> There is also the question of user documentation of the shortcut keywords for specifying packet overhead compensation. AFAIK that is still absent from the tc-cake manpage.
> 
> 	I would prefer very much if we could just rip the keywords out and just leave “overhead N” in there. As is the keywords are inconsistent in that some act as modifiers that “add” to already specified values while others replace older values whole-sale.

	Turns out that this appears to be untrue the keywords are all additive, with the exception of “overhead N” which re-sets overhead to N before adding later defined keyword_values to N… So please accept my apology for not reading the code before commenting. I believe the general points below still stand though.

Best Regards
	Sebastian


> Also these keywords do not really simplify the challenges in choosing the correct overhead at all, so my vote would be to go from today’s:
> 
>      [ atm | noatm* ] [ overhead N | conservative | raw* ]
> 
> to:
> 	[ atm | noatm* ] [ overhead N | raw* ]
> 
> 	So, instead of documenting the keywords, remove even more (and add an info/error message to tc to warn users still using the old ones). If that is not acceptable then I would propose to simply create named keywords for the individual components that make up the overhead and apply these in an additive fashion, like:
> 
> 	ppp pppoe ethernet llc snap atmpad aal5sar
> 
> which would translate into:
> 
> 	2 + 6 + 14 + 3 + 5 + 2 + 8 = 40
> 
> 	So, I propose to just take the components and their names from https://web.archive.org/web/20150606220856/http://ace-host.stuart.id.au/russell/files/tc/tc-atm/ and ad vlan, mac, and fcs to the mix (and maybe the effective additional overhead required to shape ethernet, IFG and friends). 
> 
> 	But honestly I believe we would be better off by directing users to decent information how to deduce the applicable overhead in each individual case. Say the link above and maybe a link to https://github.com/moeller0/ATM_overhead_detector might be of more help than trying to simply a complex situation?
> 	While at it, I also would remove the “conservative” keyword since with that we make a promise to our users we might not be able to keep as we have no guarantee of a maximum overhead possible.
> 
> Best Regards
> 	Sebastian
> 
>> 
>> I could work on those two points tomorrow.
>> 
>> The sch_cake kmod itself is currently at a stable point immediately before a period of planed experimentation. I recommend importing specifically commit adddf1056896f970166933d390d2300c5a00cd08 (which is HEAD right now) to preserve that.
>> 
>> - Jonathan Morton
>> 
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub, or mute the thread.
>> 
>

kdarbyshirebryant commented:
Agree with Sebastian.  Lose wash, a little frustrating as I'd had a branch with PR hanging around that did just that, recently deleted as the misfeature has previously been 'tolerated'.   Also definitely agree with removing all those extra overhead keywords.

I'm happy with being maintainer if required but Hannu, may I please have some help from you in getting the initial cake & tc patches up to date.  Ideally it needs to be easily achievable so as to be maintainable until cake goes upstream.  Sorry, my head isn't quite on top of this as I've been dealing with some router recovery issues.

hnyman commented:
I have amended the commit (and it is rebased to the current LEDE head)
* import the cake commit recommended by @chromi 
* define Kevin as the maintainer for cake, thanks @kdarbyshirebryant 

@kdarbyshirebryant , regarding maintenance:
* The cake part is easily maintainable, as the Makefile here just defines the commit to be fetched from the original cake repo and sets date as the version.
* tc patch may require some manual diff work if the later changes in tc-adv are complex. I actually created my patch a few months ago by making a diff of your tc-adv4.3 branch (that contained tc-adv applied to iproute 4.3) to plain iproute. Then I pruned the non-essential parts (e.g. extra/old cake versions still present in tc-adv, changes to other qdiscs, man pages etc.) from the patch and left only the cake stuff to be compiled. And that was not actually very much. Basically just a new module for tc, small changes to a header file and Makefile. I have since then updated the patch to include the two commits from @chromi in April.

kdarbyshirebryant commented:
To try and keep everyone in the loop this has also been sent to the cake 
list.

LR;DR - There's a pull request to the LEDE tree to integrate cake as a 
patch to tc as well as a small, separate core package for 
kmod-sched-cake.  This is a good thing.  The pull request is 
https://github.com/lede-project/source/pull/72

Hannu Nyman is the author of the patch included in the PR - I've 
volunteered to be maintainer until cake really, really, really goes 
upstream.  Lucky me :-)

Hannu's patch included an improvement in column alignment on tc -s qdisc 
show output.  I have updated Dave's tc-adv tree to include this cosmetic 
change 
https://github.com/dtaht/tc-adv/commit/54794117daef5dd16e0ccec4b821f0c67e6b9ede

Jonathan made reference to "a particular misfeature that should be 
deleted".  This mis-feature is DSCP washing.  This has been brought up 
before and to that end I put in some pull requests removing said 
facility to Dave's cake & tc-adv repos.  On that previous occasion it 
was decided it could stay but again it has been mentioned as being 
something undesirable.  I dug around, found my previous 'nosquashwash' 
trees, rebased them and have submitted pull requests again.

https://github.com/dtaht/tc-adv/pull/10
https://github.com/dtaht/sch_cake/pull/25

Maybe it will save a little work.  Or if JM happy I will commit them as 
is.  I personally would like this stumbling block removed for better or 
for worse, let's not stumble over it again :-)

Kevin





On 31/05/16 19:46, Hannu Nyman wrote:
> I have amended the commit (and it is rebased to the current LEDE head)
>
>   * import the cake commit recommended by @chromi
>     <https://github.com/chromi>
>   * define Kevin as the maintainer for cake, thanks @kdarbyshirebryant
>     <https://github.com/kdarbyshirebryant>
>
> @kdarbyshirebryant <https://github.com/kdarbyshirebryant> , regarding
> maintenance:
>
>   * The cake part is easily maintainable, as the Makefile here just
>     defines the commit to be fetched from the original cake repo and
>     sets date as the version.
>   * tc patch may require some manual diff work if the later changes in
>     tc-adv are complex. I actually created my patch a few months ago by
>     making a diff of your tc-adv4.3 branch (that contained tc-adv
>     applied to iproute 4.3) to plain iproute. Then I pruned the
>     non-essential parts (e.g. extra/old cake versions still present in
>     tc-adv, changes to other qdiscs, man pages etc.) from the patch and
>     left only the cake stuff to be compiled. And that was not actually
>     very much. Basically just a new module for tc, small changes to a
>     header file and Makefile. I have since then updated the patch to
>     include the two commits from @chromi <https://github.com/chromi> in
>     April.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lede-project/source/pull/72#issuecomment-222782884>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGMuq2vMBhtdKSJEZTkLXTVLdqomM63hks5qHIINgaJpZM4IqaPW>.
>

hnyman commented:
I have update the PR once more to reflect today's changes. 

cake: I adjusted the source commit to https://github.com/dtaht/sch_cake/commit/292b57c714f0f9a856c5822b7b3b9fb931873db7 to take in the two new cake commits (wash removal, Makefile thing)

tc:
- wash removal ( https://github.com/dtaht/tc-adv/commit/a5303c9c81187898dec96f8317d749b35d38e765 )
- additionally, @kdarbyshirebryant  has removed the unnecessary "tc_cake_old_xstats" from include/linux/pkt_sched.h in his own repo ( https://github.com/kdarbyshirebryant/tc-adv/commit/3024c5aee18a719df7520bdb0993ecd78c919291 ), so I removed that from my patch, too. 

Compile- & run-tested for ar71xx/wndr3800.

Hopefully this is now a clean and simple cake ;-)

kdarbyshirebryant commented:
Looks good to me and I'm definitely in a position to maintain.  Thanks Hannu. @nbd168 We spoke about putting cake as a base package, here it is lean and mean thanks to Hannu :-)

hnyman commented:
One more round of tinkering by the cake devs.
tc slightly modified due to https://github.com/dtaht/tc-adv/commit/8188b9d5ff281188ca993b02309d3706f260d59e and https://github.com/dtaht/tc-adv/commit/7f7add6e29af5eea304fe0830a057cda1cc9f691

blogic commented:
@jow- @nbd168 LGTM. ok to merge ?

jow- commented:
LGTM

nbd168 commented:
LGTM

KanjiMonster commented:
`sched-cake` should be in `package/kernel` since it's a kernel module.

nbd168 commented:
Yes, I overlooked that part

kdarbyshirebryant commented:
And called kmod-sched-cake perhaps?

--
Cheers,

Kevin@Darbyshire-Bryant.me.uk<mailto:Kevin@darbyshire-bryant.me.uk>
Sent from my phone, apologies for brevity, spelling & top posting

On 5 Jun 2016, at 13:31, nbd168 <notifications@github.com<mailto:notifications@github.com>> wrote:


Yes, I overlooked that part

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/lede-project/source/pull/72#issuecomment-223810597>, or mute the thread<https://github.com/notifications/unsubscribe/AGMuq8lJC1oZyO0PrdBVL9EAwoMBes9nks5qIsGRgaJpZM4IqaPW>.

kdarbyshirebryant commented:
Please see https://github.com/lede-project/source/pull/87 for V3 which addresses the Makefile move.  Has been compile & run tested on Archer C7 v2.
